### PR TITLE
Add support for datastore v1 API

### PR DESCRIPTION
**Pull request by juruen**
Google will deprecate v1beta2, the version that  is currently used by this library, on September 30th, which will make the library unusable. 

This patch adds support for the v1 API.

https://cloud.google.com/datastore/release-notes

### CODE REVIEW

File: src/main/java/com/spotify/asyncdatastoreclient/AllocateIds.java [hunk 1]
```diff
@@ -32,7 +30,7 @@
   private final List<Key> keys;
 
   AllocateIds() {
-    this.keys = Lists.newArrayList();
+    this.keys = new ArrayList<>();
```
Comment (alanw) [inline src/main/java/com/spotify/asyncdatastoreclient/AllocateIds.java]: Why change this?

File: src/main/java/com/spotify/asyncdatastoreclient/Batch.java [hunk 2]
```diff
@@ -16,19 +16,20 @@
...
 /**
  * A batch mutation statement.
  *
  * Will batch a collection of mutation operations into a single operation.
  */
-public class Batch implements MutationStatement {
+public class Batch {
 
-  private final List<MutationStatement> statements = Lists.newArrayList();
+  private final List<MutationStatement> statements = new ArrayList<>();
```
Comment (alanw) [inline src/main/java/com/spotify/asyncdatastoreclient/Batch.java]: Why change this?

File: src/main/java/com/spotify/asyncdatastoreclient/Query.java [hunk 3]
```diff
@@ -31,12 +33,12 @@
...
-  private final DatastoreV1.Query.Builder query;
-  private final List<DatastoreV1.Filter> filters;
+  private final com.google.datastore.v1.Query.Builder query;
+  private final List<Filter> filters;
 
   Query() {
-    query = DatastoreV1.Query.newBuilder();
-    filters = Lists.newArrayList();
+    query = com.google.datastore.v1.Query.newBuilder();
+    filters = new ArrayList<>();
```
Comment (alanw) [inline src/main/java/com/spotify/asyncdatastoreclient/Query.java]: Why change this?

File: src/main/java/com/spotify/asyncdatastoreclient/Entity.java [hunk 4]
```diff
@@ -413,22 +412,21 @@ public boolean equals(final Object obj) {
...
-        }).collect(Collectors.toList());
-    final DatastoreV1.Entity.Builder prepared = DatastoreV1.Entity.newBuilder(entity)
-            .setKey(getKey().getPb(namespace));
-    return prepared.clearProperty().addAllProperty(propertiesLocal).build();
+  com.google.datastore.v1.Entity getPb(final String namespace) {
+    final Map<String, com.google.datastore.v1.Value> propertiesLocal = entity.getProperties()
+      .entrySet()
+      .stream()
+      .collect(Collectors.toMap(
+        Map.Entry::getKey, e -> Value.builder(e.getValue()).build().getPb(namespace)));
```
Comment (alanw) [inline src/main/java/com/spotify/asyncdatastoreclient/Entity.java]: Not a biggie. Should be indented by 4.

File: src/main/java/com/spotify/asyncdatastoreclient/Entity.java [hunk 5]
```diff
@@ -413,22 +412,21 @@ public boolean equals(final Object obj) {
...
+
+    return com.google.datastore.v1.Entity
+      .newBuilder()
```
Comment (alanw) [inline src/main/java/com/spotify/asyncdatastoreclient/Entity.java]: Not a biggie. Should be indented by 4.

File: src/main/java/com/spotify/asyncdatastoreclient/Filter.java [hunk 6]
```diff
@@ -45,32 +43,32 @@ public Filter(final String name, final Operator op, final Value value) {
     this.value = value;
   }
 
-  DatastoreV1.Filter getPb() {
-    final DatastoreV1.PropertyFilter.Builder filter = DatastoreV1.PropertyFilter.newBuilder()
-        .setProperty(DatastoreV1.PropertyReference.newBuilder().setName(name))
-        .setValue(value.getPb());
+  com.google.datastore.v1.Filter getPb(String namespace) {
```
Comment (alanw) [inline src/main/java/com/spotify/asyncdatastoreclient/Filter.java]: final String namespace`

File: src/main/java/com/spotify/asyncdatastoreclient/Key.java [hunk 7]
```diff
@@ -379,16 +378,27 @@ public boolean equals(final Object obj) {
...
-              .setPartitionId(DatastoreV1.PartitionId.newBuilder().setNamespace(namespace)).build();
+      return com.google.datastore.v1.Key
+          .newBuilder(key)
+          .setPartitionId(
+              com.google.datastore.v1.PartitionId
+                  .newBuilder().setNamespaceId(namespace)).build();
     }
   }
+
+  private static boolean isId(com.google.datastore.v1.Key.PathElement element) {
```
Comment (alanw) [inline src/main/java/com/spotify/asyncdatastoreclient/Key.java]: `final`

File: src/main/java/com/spotify/asyncdatastoreclient/Key.java [hunk 8]
```diff
@@ -379,16 +378,27 @@ public boolean equals(final Object obj) {
...
+    return element.getIdTypeCase() == com.google.datastore.v1.Key.PathElement.IdTypeCase.ID;
+  }
+
+  private static boolean isName(com.google.datastore.v1.Key.PathElement element) {
```
Comment (alanw) [inline src/main/java/com/spotify/asyncdatastoreclient/Key.java]: `final`

File: src/main/java/com/spotify/asyncdatastoreclient/MutationStatement.java [hunk 9]
```diff
@@ -25,6 +25,6 @@
  */
 public interface MutationStatement extends Statement {
 
-  DatastoreV1.Mutation getPb(String namespace);
+  Mutation getPb(String namespace);
```
Comment (alanw) [inline src/main/java/com/spotify/asyncdatastoreclient/MutationStatement.java]: `final`

File: src/main/java/com/spotify/asyncdatastoreclient/Query.java [hunk 10]
```diff
@@ -46,8 +48,10 @@
    * @return this query statement.
    */
   public Query keysOnly() {
-    query.addProjection(DatastoreV1.PropertyExpression.newBuilder()
-                            .setProperty(DatastoreV1.PropertyReference.newBuilder().setName("__key__")));
+    query.addProjection(
```
Comment (alanw) [inline src/main/java/com/spotify/asyncdatastoreclient/Query.java]: Not a biggie. Should be indented by 4.

File: src/main/java/com/spotify/asyncdatastoreclient/Value.java [hunk 11]
```diff
@@ -116,9 +122,16 @@ public Builder value(final Object value) {
      * that are not recognised.
      */
     public Builder value(final List<Object> values) {
-      this.value.addAllListValue(values.stream()
-                                     .map(valueLocal -> Value.builder(valueLocal).build().getPb())
-                                     .collect(Collectors.toList()));
+      final ArrayValue arrayValues = ArrayValue
```
Comment (alanw) [inline src/main/java/com/spotify/asyncdatastoreclient/Value.java]: Not a biggie. Should be indented by 4.

File: src/main/java/com/spotify/asyncdatastoreclient/Value.java [hunk 12]
```diff
@@ -129,9 +142,19 @@ public Builder value(final List<Object> values) {
...
+      this.excludeFromIndexes = !indexed;
       return this;
     }
+
+  }
+
+  private static Timestamp toTimestamp(Date date) {
+    final long millis = date.getTime();
+    return Timestamp
+      .newBuilder()
```
Comment (alanw) [inline src/main/java/com/spotify/asyncdatastoreclient/Value.java]: Not a biggie. Should be indented by 4.

File: src/main/java/com/spotify/asyncdatastoreclient/Value.java [hunk 13]
```diff
@@ -567,16 +581,19 @@ public boolean isList() {
...
     return value;
   }
 
-  DatastoreV1.Value getPb(final String namespace) {
-    if (value.hasKeyValue()) {
-      return DatastoreV1.Value.newBuilder(value).setKeyValue(getKey().getPb(namespace)).build();
+  com.google.datastore.v1.Value getPb(final String namespace) {
+    final com.google.datastore.v1.Key key = value.getKeyValue();
+    if (key.getPathCount() > 0) {
+      return com.google.datastore.v1.Value.
```
Comment (alanw) [inline src/main/java/com/spotify/asyncdatastoreclient/Value.java]: Not a biggie. Should be indented by 4.

File: src/main/java/com/spotify/asyncdatastoreclient/Value.java [hunk 14]
```diff
@@ -593,27 +610,40 @@ public boolean equals(final Object obj) {
...
+        return String.valueOf(value.getIntegerValue());
+      case DOUBLE_VALUE:
+        return String.valueOf(value.getDoubleValue());
+      case BOOLEAN_VALUE:
+        return String.valueOf(value.getBooleanValue());
+      case ENTITY_VALUE:
+        return getEntity().toString();
+      case ARRAY_VALUE:
+        return "[" +
+          getList()
```
Comment (alanw) [inline src/main/java/com/spotify/asyncdatastoreclient/Value.java]: Not a biggie. Should be indented by 4.

File: src/main/java/com/spotify/asyncdatastoreclient/Value.java [hunk 15]
```diff
@@ -593,27 +610,40 @@ public boolean equals(final Object obj) {
...
+            .collect(Collectors.joining(", ")) + "]";
+      case NULL_VALUE:
+        return "<null>";
+      default:
+        return value.toString();
     }
-    return value.toString();
+  }
+
+  private Date toDate(Timestamp timestamp) {
```
Comment (alanw) [inline src/main/java/com/spotify/asyncdatastoreclient/Value.java]: `final`

File: src/main/java/com/spotify/asyncdatastoreclient/Value.java [hunk 16]
```diff
@@ -593,27 +610,40 @@ public boolean equals(final Object obj) {
...
-      return "<binary>";
-    } else if (value.hasEntityValue()) {
-      return getEntity().toString();
-    } else if (value.hasKeyValue()) {
-      return getKey().toString();
-    } else if (value.getListValueCount() > 0) {
-      return "[" + getList().stream()
-          .map(Value::toString)
-          .collect(Collectors.joining(", ")) + "]";
+    final com.google.datastore.v1.Value.ValueTypeCase c = value.getValueTypeCase();
```
Comment (alanw) [inline src/main/java/com/spotify/asyncdatastoreclient/Value.java]: `switch(value.getValueTypeCase())`

File: src/test/java/com/spotify/asyncdatastoreclient/InsertTest.java [hunk 17]
```diff
@@ -22,18 +22,14 @@
...
 
 import java.util.Date;
 import java.util.List;
 import java.util.Random;
 
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertTrue;
-import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.fail;
+import static org.junit.Assert.*;
```
Comment (alanw) [inline src/test/java/com/spotify/asyncdatastoreclient/InsertTest.java]: I prefer to explicitly import and not wildcard import.

File: src/test/java/com/spotify/asyncdatastoreclient/KeyTest.java [hunk 18]
```diff
@@ -16,17 +16,11 @@
...
 import org.junit.Test;
 
 import java.util.List;
 
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertNotNull;
-import static org.junit.Assert.assertNull;
-import static org.junit.Assert.assertTrue;
-import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.*;
```
Comment (alanw) [inline src/test/java/com/spotify/asyncdatastoreclient/KeyTest.java]: I prefer to explicitly import and not wildcard import.

File: src/test/java/com/spotify/asyncdatastoreclient/QueryTest.java [hunk 19]
```diff
@@ -326,6 +326,7 @@ public void testQueryDateFilter() throws Exception {
     assertEquals(10, entities.size());
   }
 
+
```
Comment (alanw) [inline src/test/java/com/spotify/asyncdatastoreclient/QueryTest.java]: Unnecessary newline.

File: src/main/java/com/spotify/asyncdatastoreclient/AllocateIds.java (hunk 1 above)
Comment (juruen) [inline src/main/java/com/spotify/asyncdatastoreclient/AllocateIds.java]: I was initially trying to reduce dependencies. My bad. I'll change it back :)

File: src/main/java/com/spotify/asyncdatastoreclient/Batch.java (hunk 2 above)
Comment (juruen) [inline src/main/java/com/spotify/asyncdatastoreclient/Batch.java]: Same here :)

File: src/main/java/com/spotify/asyncdatastoreclient/Query.java (hunk 3 above)
Comment (juruen) [inline src/main/java/com/spotify/asyncdatastoreclient/Query.java]: Same here.

File: src/main/java/com/spotify/asyncdatastoreclient/Entity.java (hunk 4 above)
Comment (juruen) [inline src/main/java/com/spotify/asyncdatastoreclient/Entity.java]: ACK!

File: src/main/java/com/spotify/asyncdatastoreclient/Entity.java (hunk 5 above)
Comment (juruen) [inline src/main/java/com/spotify/asyncdatastoreclient/Entity.java]: ACK!

File: src/main/java/com/spotify/asyncdatastoreclient/Filter.java (hunk 6 above)
Comment (juruen) [inline src/main/java/com/spotify/asyncdatastoreclient/Filter.java]: ACK!

File: src/main/java/com/spotify/asyncdatastoreclient/Key.java (hunk 7 above)
Comment (juruen) [inline src/main/java/com/spotify/asyncdatastoreclient/Key.java]: ACK!

File: src/main/java/com/spotify/asyncdatastoreclient/Key.java (hunk 8 above)
Comment (juruen) [inline src/main/java/com/spotify/asyncdatastoreclient/Key.java]: ACK!

File: src/main/java/com/spotify/asyncdatastoreclient/MutationStatement.java (hunk 9 above)
Comment (juruen) [inline src/main/java/com/spotify/asyncdatastoreclient/MutationStatement.java]: ACK!

File: src/main/java/com/spotify/asyncdatastoreclient/Query.java (hunk 10 above)
Comment (juruen) [inline src/main/java/com/spotify/asyncdatastoreclient/Query.java]: ACK!

File: src/main/java/com/spotify/asyncdatastoreclient/Value.java (hunk 11 above)
Comment (juruen) [inline src/main/java/com/spotify/asyncdatastoreclient/Value.java]: ACK!

File: src/main/java/com/spotify/asyncdatastoreclient/Value.java (hunk 12 above)
Comment (juruen) [inline src/main/java/com/spotify/asyncdatastoreclient/Value.java]: ACK!

File: src/main/java/com/spotify/asyncdatastoreclient/Value.java (hunk 13 above)
Comment (juruen) [inline src/main/java/com/spotify/asyncdatastoreclient/Value.java]: ACK!

File: src/main/java/com/spotify/asyncdatastoreclient/Value.java (hunk 14 above)
Comment (juruen) [inline src/main/java/com/spotify/asyncdatastoreclient/Value.java]: ACK!

File: src/main/java/com/spotify/asyncdatastoreclient/Value.java (hunk 15 above)
Comment (juruen) [inline src/main/java/com/spotify/asyncdatastoreclient/Value.java]: ACK!

File: src/main/java/com/spotify/asyncdatastoreclient/Value.java (hunk 16 above)
Comment (juruen) [inline src/main/java/com/spotify/asyncdatastoreclient/Value.java]: ACK!

File: src/test/java/com/spotify/asyncdatastoreclient/InsertTest.java (hunk 17 above)
Comment (juruen) [inline src/test/java/com/spotify/asyncdatastoreclient/InsertTest.java]: No problem. I just changed my IDEA configuration to do this.

File: src/test/java/com/spotify/asyncdatastoreclient/KeyTest.java (hunk 18 above)
Comment (juruen) [inline src/test/java/com/spotify/asyncdatastoreclient/KeyTest.java]: Same as above.

File: src/test/java/com/spotify/asyncdatastoreclient/QueryTest.java (hunk 19 above)
Comment (juruen) [inline src/test/java/com/spotify/asyncdatastoreclient/QueryTest.java]: Yup!

**Comment by juruen** [comment]
@alanw Just updated the PR to fix your comments.

**Pull request merged**